### ISSUE TITLE
Random fails update_imports_on_move

### ISSUE THREAD

**Issue by ggsalas** [jose-elias-alvarez/nvim-lsp-ts-utils#27]
I have configured nvim-lsp-ts-utils to only use `update_imports_on_move`. Works well but sometimes fails. I'm using [dirvish](https://github.com/justinmk/vim-dirvish) so I have tried with `mv` and with `git mv` commands. 

I'm not sure how to report the bug, because when fails I see no imports updated but without any error. There is a log or something I can send to debug?

This is my config: 
```lua
lsp.tsserver.setup({
    on_attach = function(client)
        client.resolved_capabilities.document_formatting = false

        local ts_utils = require("nvim-lsp-ts-utils")

        ts_utils.setup {
            debug = true,
            disable_commands = true,
            enable_import_on_completion = false,
            eslint_enable_code_actions = false,
            eslint_enable_disable_comments = false,
            eslint_enable_diagnostics = false,
            enable_formatting = false,
            format_on_save = false,
            complete_parens = false,
            signature_help_in_parens = false,

            update_imports_on_move = true,
            require_confirmation_on_move = false
        }

        ts_utils.setup_client(client)
    end
})
```

**Comment by jose-elias-alvarez** [comment]
Moving files from within Vim is a little finicky if a non-standard buffer (i.e. a buffer that contains anything other than a file) is focused. The plugin works around this by temporarily focusing one of the files being moved, but it's not a perfect solution and needs more testing. 

With `debug` enabled, `:messages` should show you a series of messages when you move a file confirming that the watcher is attempting to update imports, followed by the source and the target destinations. The next time imports don't get updated, could you confirm whether you see those messages?

**Comment by jose-elias-alvarez** [comment]
Also: if you are only interested in updating imports on move, you can compress your config pretty significantly, since almost everything is disabled by default:

```lua
lsp.tsserver.setup({
    on_attach = function()
        require("nvim-lsp-ts-utils").setup {
            debug = true,
            disable_commands = true,
            update_imports_on_move = true
        }
        -- not calling setup_client will prevent all ESLint and formatter integration
    end
})
```

**Comment by jose-elias-alvarez** [comment]
There are also some changes on the `develop` branch of the plugin that deal with the non-standard buffer issue a little more reliably and also simplify the logic that checks whether the plugin thinks a file was moved, if you're still having problems and want to give it a shot. You shouldn't have any incompatibilities – the only change that will affect your use case is that [plenary.nvim](https://github.com/nvim-lua/plenary.nvim) is required.

**Comment by ggsalas** [comment]
1. I have moved to the develop branch on `~/.config/nvim/plugged/nvim-lsp-ts-utils` and I'm using `nvim-lua/plenary.nvim` plugin.

Now, if I move a file or a folder from the command line (with nvim open on other window) I not see any changes. Seems previously was working with this method for the files. For the folders I see this error ```RPC[Error] code_name = MethodNotFound, message = "method workspace/executeCommand is not supported by 
any of the servers registered for the current buffer" ```

2. Could be that I need to open a ts to initiate the watcher?

3. Maybe is easy to create a command to rename and move files and folders. I have tried `lua require'nvim-lsp-ts-utils'.rename_file()` and works well for the current open file. I would like to see if there is a way to specify one or many files and if there is a way to use to move a folder.

Thanks for your help @jose-elias-alvarez.

**Comment by jose-elias-alvarez** [comment]
Hmm, that's strange. Both the `develop` and `main` branches require you to open a file that `tsserver` handles (`ts`, `tsx`, `js`, or `jsx`) n order for the watcher to run, since (as you can see from the error message) the language server must be open to handle the request. 

Are you doing anything special to start the watcher (starting it manually, perhaps)? I can't replicate what you're seeing on my end, but yes, the watcher does essentially require a TypeScript file to be open. Does moving / renaming files work as expected if you have a file open?

It's possible to expand the manual `rename_file` method (and I'll add the ability for it to handle folders for consistency), but for multiple files I'd prefer to keep working on improving the watcher functionality, since I think it's a lot more comfortable for people to move files with whatever method they are already used to. Thank you for your patience and help!

**Comment by ggsalas** [comment]
Hi @jose-elias-alvarez, You're right, for this to work I need to have open a js file at the same time I run the move command on the console. For example, if I open a js file, then open a md file and then run `mv folder folder1` I will see this error ```RPC[Error] code_name = MethodNotFound, message = "method workspace/executeCommand is not supported by any of the servers registered for the current buffer"```

Now makes sense to me. But  In my opinion that rule is easy to forget (to have a ts file open). If the listener is related with file movements maybe must should not require an specific buffer opened.

With Dirvish:
I have tried again on develop branch and Dirvish and it works! Previously I was used dirvish without open a file, that was my error. But sometimes fails:

- 1. open a ts file, then dirvish, then `:!mv file.tsx file1.tsx` works well
- 2. open a ts file, then dirvish, then `:!mv folder folder1` I get this error ```RPC[Error] code_name = MethodNotFound, message = "method workspace/executeCommand is not supported by any of the servers registered for the current buffer"``` 

I think this is an error with the scss imports. If I remove `import "./style.scss"` from the tsx file (inside the folder) I get no error on move the file.

Also I get no error if I run same command on console (while I have nvim with a ts file open)

But If I move a single scss file `:!mv style.scss style1.scss` the import on the tsx file not changes (`import "./style.scss"`) no matters If I do this on the console or on Dirvish.

**Comment by ggsalas** [comment]
Hi again, I have detected that the imports update fails for some folders. I'm using [this example repo](https://github.com/carlosvillu/pajarito):
1. open `src/index.js`
2. run this command on the console `mv src/components src/componentsGG`. 

The result is that not any import has been updated.
if I run `: messages` I get this:
```
watching directory /Users/ggsalas/developer/pajarito/src                                                                                                     
attempting to update imports                                                                                                                                 
source: /Users/ggsalas/developer/pajarito/src/components                                                                                                     
target: /Users/ggsalas/developer/pajarito/src/componentsGG
```

Sorry for not be more helpfull. I think this is an awesome feature!

**Comment by jose-elias-alvarez** [comment]
> For example, if I open a js file, then open a md file and then run mv folder folder1 I will see this error RPC[Error] code_name = MethodNotFound, message = "method workspace/executeCommand is not supported by any of the servers registered for the current buffer"

Ahh, okay, I think I understand the issue now. I just pushed a commit to the `develop` branch that should fix this case, so could you give it a try when you have a chance? You'll still have to open a `ts` / `js` file to start the watcher, but that's because the file moving functionality fundamentally depends on `tsserver`, as I mentioned. 

> But If I move a single scss file :!mv style.scss style1.scss the import on the tsx file not changes (import "./style.scss") no matters If I do this on the console or on Dirvish. 

`.scss` files aren't handled by `tsserver`, so there's no way for the language server to update these types of imports (even VS Code can't do it with default `tsserver` functionality, though there may be a plugin out there). The plugin's role is to watch the directory and send a command to the language server in response to events, so even if I tell it to react to `scss` file change events, the server won't do anything in response.

>     1. open `src/index.js`
> 
>     2. run this command on the console `mv src/components src/componentsGG`.

Wow, thanks for the example repo. I was able to see that my mechanism to handle directories wasn't handling nested directories properly, so I was able to improve that. The other thing about this example is that `index.js` doesn't import anything from `components/`, so it's possible that the buffers are loaded but aren't visible. If you're still having issues on the latest `develop` commit, could you verify if you see the relevant buffers when you run the Vim command `:ls`?

Thank you for all of your help improving this feature!

**Comment by ggsalas** [comment]
> Ahh, okay, I think I understand the issue now. I just pushed a commit to the develop branch that should fix this case, so could you give it a try when you have a chance? You'll still have to open a ts / js file to start the watcher, but that's because the file moving functionality fundamentally depends on tsserver, as I mentioned.

This is working perfect on my end 👍 
Do you know if I can start tsserver if nvim is opened inside a folder that has a `.js`, `.ts` or `tsconfig.json` file? I think that would be best for my workflow, but I understand is not high prority.

> .scss files aren't handled by tsserver, so there's no way for the language server to update these types of imports (even VS Code can't do it with default tsserver functionality, though there may be a plugin out there). The plugin's role is to watch the directory and send a command to the language server in response to events, so even if I tell it to react to scss file change events, the server won't do anything in response.

I understand, makes sense. The issue is that the scss errors breaks the complete functionality. Maybe there is a way to show some warning, but fix all the js and ts imports. Also on the example repo works well because it uses scss modules. If the project uses scss files it fails and do nothing. You can try this bug like this:
1. create a new repo with `npx create-react-app example-imports`
2. open a js file with nvim
2. create a new folder `mkdir src/App `
3. Move the App css and js files into the new folder
  `mv ./src/App.js ./src/App/App.js` the js imports are fixed correctly
  `mv ./src/App.css ./src/App/App.css`
4. Rename the folder: `mv ./src/App ./src/AppGG` 
  At this point you should see this error message and **no imports fixed** (on `src/index.js` the `App.js` import should be uptated to `import App from "./AppGG/App"` but is not). `RPC[Error] code_name = MethodNotFound, message = "method workspace/executeCommand is not supported by any of the serv
ers registered for the current buffer"`

> Wow, thanks for the example repo. I was able to see that my mechanism to handle directories wasn't handling nested directories properly, so I was able to improve that, and it seems to be working on my end now.
> 
> The other thing about this example is that index.js doesn't import anything from components/, so it's possible that the buffers are loaded but aren't visible. If you're still having issues on the latest develop commit, could you verify if you see the relevant buffers when you run the Vim command :ls?

This is working on my end too 👍 . I have tested from Dirvish and from the console.

**Comment by jose-elias-alvarez** [comment]
Great to know that the changes helped! 

I followed your instructions (and again, thanks for the easily reproducible setup), and interestingly, I don't see the same error. Are you using a language server that's active for CSS files, by any chance?

Either way, I believe that the cause is an issue in the filetype pattern match I wrote, which caused the watcher to attempt to update imports on `App.css`. I just pushed a change that should take care of it, so let me know!

**Comment by ggsalas** [comment]
Yes, sorry I forgot to mention I have installed css LSP (is useful to have autocompletion) and my config is only this: `
lsp.css.setup {}
`

I have tested  on develop branch and I not see any error with the css 👍 Amazing, thanks!

Thanks for your suggestion about start tsserver on VimEnter. I think you're right that requires some manual language server management and could have some edge cases. I will trie. Thanks again!
